File: typescript-driver/src/postgres-source.ts
// Lookup + descarga del driver TypeScript desde data.ts_cliente_driver via psql.
// Usa subprocess (sin dep node-postgres) — espejo del patrón de actualiza-cliente-*.

import { spawnSync } from 'node:child_process';

export interface DriverMeta {
  version: number;
  entorno: string;
  hashSha256: string;
  tamano: number;
}

export interface PgConn {
  host: string;
  port: number;
  database: string;
  user: string;
  password: string;
}

export class PostgresDriverSource {
  constructor(
    private readonly conn: PgConn,
    public readonly entorno: 'test' | 'produccion',
  ) {
    if (entorno !== 'test' && entorno !== 'produccion') {
      throw new Error(`entorno inválido: ${entorno} (debe ser test|produccion)`);
    }
  }

  private psql(sql: string): string {
    const r = spawnSync(
      'psql',
      ['-h', this.conn.host, '-p', String(this.conn.port), '-U', this.conn.user,
        '-d', this.conn.database, '-v', 'ON_ERROR_STOP=1', '-t', '-A', '-X', '-q'],
      { input: sql, env: { ...process.env, PGPASSWORD: this.conn.password }, encoding: 'utf-8' },
    );
    if (r.status !== 0) throw new Error(`psql falló: ${r.stderr.trim()}`);
    return r.stdout.trim();
  }

  async lookup(): Promise<DriverMeta | null> {
    const sql = `SELECT (row_to_json(x.*))::text FROM (
      SELECT ok, message, data
      FROM fn.ts_cliente_driver_lookup(jsonb_build_object(
        'session', jsonb_build_object('trusted', true),
        'entorno', '${this.entorno}'
      ))
    ) x;`;
    const raw = this.psql(sql);
    const row = JSON.parse(raw) as { ok: boolean; message: string; data: any };
    if (!row.ok) {
      if (row.message === 'ts_cliente_driver.no_disponible') return null;
      throw new Error(`lookup falló: ${row.message}`);
    }
    return {
      version: row.data.version,
      entorno: row.data.entorno,
      hashSha256: row.data.hash_sha256,
      tamano: row.data.tamano,
    };
  }

  async descargar(version?: number): Promise<Uint8Array> {
    const verLit = version === undefined ? "''" : `'${version}'`;
    const sql = `SELECT (row_to_json(x.*))::text FROM (
      SELECT ok, message, data
      FROM fn.ts_cliente_driver_descargar(jsonb_build_object(
        'session', jsonb_build_object('trusted', true),
        'entorno', '${this.entorno}',
        'version', ${verLit}
      ))
    ) x;`;
    const raw = this.psql(sql);
    const row = JSON.parse(raw) as { ok: boolean; message: string; data: any };
    if (!row.ok) throw new Error(`descarga falló: ${row.message}`);
    return Buffer.from(row.data.bytes_b64, 'base64');
  }

  async publicar(bytes: Uint8Array, hashSha256: string, notas?: string): Promise<{ id: number; version: number }> {
    const b64 = Buffer.from(bytes).toString('base64');
    const notasFrag = notas ? `, 'notas', '${notas.replace(/'/g, "''")}'` : '';
    // Bytes via dollar-quoted string para evitar escapes
    const sql = `SELECT (row_to_json(x.*))::text FROM (
      SELECT ok, message, data
      FROM fn.ts_cliente_driver_publicar(jsonb_build_object(
        'session',     jsonb_build_object('trusted', true),
        'entorno',     '${this.entorno}',
        'hash_sha256', '${hashSha256}'
        ${notasFrag},
        'bytes_b64',   $CHALONA_BYTES$${b64}$CHALONA_BYTES$
      ))
    ) x;`;
    const raw = this.psql(sql);
    const row = JSON.parse(raw) as { ok: boolean; message: string; data: any };
    if (!row.ok) throw new Error(`publicar falló: ${row.message}`);
    return { id: row.data.id, version: row.data.version };
  }
}

File: dart-driver/lib/postgres_source.dart
// PostgresDriverSource: lookup + descarga del driver Dart desde data.dart_cliente_driver.
// Reemplaza ServidorSimulado con conexión real.
//
// Uso típico:
//   final source = PostgresDriverSource(
//     host: 'localhost', port: 5433, database: 'test',
//     username: 'pedro', password: 'camila',
//     entorno: 'test',
//   );
//   final meta = await source.lookup();
//   if (meta.version != driverActivo?.version) {
//     final bytes = await source.descargar(meta.version);
//     // ... cargar en runtime
//   }

import 'dart:convert';
import 'dart:typed_data';

import 'package:postgres/postgres.dart';

class DriverMeta {
  final int version;
  final String entorno;
  final String hashSha256;
  final int tamano;

  DriverMeta({
    required this.version,
    required this.entorno,
    required this.hashSha256,
    required this.tamano,
  });

  factory DriverMeta.fromJson(Map<String, Object?> j) => DriverMeta(
        version: (j['version'] as num).toInt(),
        entorno: j['entorno'] as String,
        hashSha256: j['hash_sha256'] as String,
        tamano: (j['tamano'] as num).toInt(),
      );

  @override
  String toString() =>
      'DriverMeta(v=$version entorno=$entorno tam=$tamano sha=${hashSha256.substring(0, 12)}...)';
}

class PostgresDriverSource {
  final String host;
  final int port;
  final String database;
  final String username;
  final String password;
  final String entorno;

  Connection? _conn;

  PostgresDriverSource({
    required this.host,
    required this.port,
    required this.database,
    required this.username,
    required this.password,
    required this.entorno,
  }) {
    if (entorno != 'test' && entorno != 'produccion') {
      throw ArgumentError('entorno debe ser test o produccion');
    }
  }

  Future<Connection> _connect() async {
    if (_conn != null) return _conn!;
    _conn = await Connection.open(
      Endpoint(
        host: host,
        port: port,
        database: database,
        username: username,
        password: password,
      ),
      settings: const ConnectionSettings(sslMode: SslMode.disable),
    );
    return _conn!;
  }

  Future<void> close() async {
    await _conn?.close();
    _conn = null;
  }

  /// Llamada barata: devuelve metadata (versión, hash, tamaño). Sin bytes.
  Future<DriverMeta?> lookup() async {
    final c = await _connect();
    final r = await c.execute(
      r"""
      SELECT ok, message, data::text
      FROM fn.dart_cliente_driver_lookup(jsonb_build_object(
        'session', jsonb_build_object('trusted', true),
        'entorno', $1::text
      ))
      """,
      parameters: [entorno],
    );
    final row = r.first;
    final ok = row[0] as bool;
    final message = row[1] as String;
    final dataJson = row[2] as String;
    if (!ok) {
      if (message == 'dart_cliente_driver.no_disponible') return null;
      throw StateError('lookup falló: $message');
    }
    final data = jsonDecode(dataJson) as Map<String, Object?>;
    return DriverMeta.fromJson(data);
  }

  /// Llamada cara: descarga bytes de la versión activa (o específica).
  Future<Uint8List> descargar({int? version}) async {
    final c = await _connect();
    final r = await c.execute(
      r"""
      SELECT ok, message, data::text
      FROM fn.dart_cliente_driver_descargar(jsonb_build_object(
        'session', jsonb_build_object('trusted', true),
        'entorno', $1::text,
        'version', $2::text
      ))
      """,
      parameters: [entorno, version?.toString() ?? ''],
    );
    final row = r.first;
    final ok = row[0] as bool;
    final message = row[1] as String;
    final dataJson = row[2] as String;
    if (!ok) throw StateError('descarga falló: $message');
    final data = jsonDecode(dataJson) as Map<String, Object?>;
    // Postgres encode(..., 'base64') intercala '\n' cada 76 chars
    final b64 = (data['bytes_b64'] as String).replaceAll(RegExp(r'\s'), '');
    return base64Decode(b64);
  }
}
